yt/yt/client/logging/dynamic_table_log_writer.cpp

```diff
@@ -0,0 +1,373 @@
+                YT_LOG_ERROR(ex, "Error flushing log events to dynamic table");
+            }
+
+            if (backoff.Next()) {
+                TDelayedExecutor::WaitForDuration(backoff.GetBackoff());
+            } else {
+                YT_LOG_ERROR("Flush retries exhausted, dropping log events (Size: %v)", ysonRows.size());
+                break;
+            }
+        }
+    }
+
+    i64 DoFlushIteration(const TString& ysonRows)
+    {
+        VERIFY_INVOKER_AFFINITY(Invoker_);
+
+        auto client = ClientHolder_->Client.Load();
+
+        if (!client) {
+            THROW_ERROR_EXCEPTION("Cannot flush rows without configured YT client");
+        }
+
+        auto tableMountCache = client->GetTableMountCache();
+        auto tableInfo = WaitForFast(tableMountCache->GetTableInfo(Config_->TablePath))
+            .ValueOrThrow();
+
+        tableInfo->ValidateDynamic();
+
+        TBuildingValueConsumer valueConsumer(
+            tableInfo->Schemas[ETableSchemaKind::Write],
+            TLogger{},
+            Config_->InsertRowsFormat->EnableNullToYsonEntityConversion,
+            Config_->TypeConversion);
+        // This seems like a reasonable configuration.
+        valueConsumer.SetAggregate(false);
+        valueConsumer.SetTreatMissingAsNull(true);
+        valueConsumer.SetAllowMissingKeyColumns(false);
+
+        struct TDynamicTableLogWriterBufferTag
+        { };
+
+        TStringInput inputStream(ysonRows);
+
+        TTableOutput output(CreateParserForFormat(
+            TFormat(EFormatType::Yson),
+            &valueConsumer));
+
+        PipeInputToOutput(&inputStream, &output, 64_KB);
+        auto rowBuffer = New<TRowBuffer>(TDynamicTableLogWriterBufferTag());
+        auto capturedRows = rowBuffer->CaptureRows(valueConsumer.GetRows());
+        auto rowRange = MakeSharedRange(
+            std::vector<TUnversionedRow>(capturedRows.begin(), capturedRows.end()),
+            std::move(rowBuffer));
+        auto dataWeight = GetDataWeight(rowRange);
+
+        auto transaction = WaitFor(client->StartTransaction(ETransactionType::Tablet))
+            .ValueOrThrow();
+        transaction->WriteRows(
+            Config_->TablePath,
+            valueConsumer.GetNameTable(),
+            std::move(rowRange));
+
+        WaitFor(transaction->Commit())
+            .ThrowOnError();
+
+        return dataWeight;
+    }
+
+    // TODO(achulkov2): A lot of the logic above should be separated into some
+    // sort of buffered dynamic table writer. It should be responsible for buffering
+    // rows, low/high watermarks, and performing the actual YT writes with retries.
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+class TDynamicTableLogWriterFactory
+    : public IDynamicTableLogWriterFactory
+{
+public:
+    TDynamicTableLogWriterFactory()
+        : ClientHolder_(New<TClientHolder>())
+        , WriteThreadPool_(CreateThreadPool(1, "LogYTWrite"))
```

Comment:
Do we need a thread pool here?

Reply:
There is some serialization/deserialization load, but I guess we can live with a single action queue thread for now. If necessary, we can change it to a thread pool with a configurable number of threads.